### PR TITLE
Keyboard gui backend

### PR DESCRIPTION
**Pull request by owen927**
what's in this pull request:
ringData class + keyboardFileReader
updated myoTranslationFilter - now the calculation angle of myo for keySelect in done in the filter
updated SCDDigester - MidasMain now reads keyboard setup file, passes the data into SCDDigester
depends on user input(click/ hold )SCDDigester uses keySelectAngle and
kybdGUISel to determine which char to output to send
the logics for determining the keySelectAngle will still need more work

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -62,3 +62,6 @@ PMMain/Midas/MidasGUI/GeneratedFiles/
 
 
 
+PMMain/Midas/MidasGUI/MidasGUI.vcxproj.user
+PMMain/Midas/MidasGUI/MidasGUI.vcxproj.filters
+PMMain/Midas/MidasGUI/MidasGUI.vcxproj
```
Comment (jordenh) [inline .gitignore]: Not needed. Please remove this change. *.vcxproj.user is already in the file, and others shouldnt be added.

File: PMMain/Midas/MidasGUI/MidasGUI.vcxproj.user [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    <LocalDebuggerEnvironment>PATH=$(QTDIR)\bin%3b"$(QTDIR)\bin%3b$(PATH)</LocalDebuggerEnvironment>
+
+    <QTDIR>C:\Qt\5.4\msvc2013_opengl</QTDIR>
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|Win32'">
+    <LocalDebuggerEnvironment>PATH=$(QTDIR)\bin%3b"$(QTDIR)\bin%3b$(PATH)</LocalDebuggerEnvironment>
+    <QTDIR>C:\Qt\5.4\msvc2013_opengl</QTDIR>
+
+  </PropertyGroup>
+</Project>
```
Comment (jordenh) [inline PMMain/Midas/MidasGUI/MidasGUI.vcxproj.user]: Don't commit this file. (use git rm, but make sure to back it up first)

File: PMMain/Midas/MidasGUI/MidasMain.cpp [hunk 3]
```diff
@@ -80,11 +82,12 @@ int midasMain(MidasThread *threadHandle, SequenceDisplayer* sequenceDisplayer, I
     MyoDevice* myoDevice = new MyoDevice(&sharedData, &controlState, "com.midas.midas-test", sequenceDisplayer, infoIndicator);
     MouseCtrl* mouseCtrl = new MouseCtrl();
     KybrdCtrl* kybrdCtrl = new KybrdCtrl();
+    vector<ringData> kybrdRingData = threadHandle->readKeyboardSetupFile();
```
Comment (jordenh) [inline PMMain/Midas/MidasGUI/MidasMain.cpp]: Why is this calling the function from threadHandle? It shouldnt need to be a function of MidasThread

File: PMMain/Midas/MidasGUI/MidasCommon.h [hunk 4]
```diff
@@ -173,6 +173,11 @@ struct point {
     point(int xVal = 0, int yVal = 0) : x(xVal), y(yVal) { }
 };
 
+struct ori_data {
```
Comment (lordy1992) [inline PMMain/Midas/MidasGUI/MidasCommon.h]: I don't particularly like this name. I assume ori is supposed to stand for orientation, but the only reason I can figure that out is because I see that roll, pitch and yaw are members of the struct. The name should be more standalone.

File: PMMain/Midas/MidasGUI/MidasThread.cpp [hunk 5]
```diff
@@ -41,3 +42,97 @@ void MidasThread::threadEmitStateString(std::string str)
...
+                    std::cout << "loading!!!!!\n";
+                }
+
+            }
+
+            if (ringInOutSwitch == 0)
+            {
+                std::cout << "Ring In Vector!!!!!\n";
+                ringTemp.setRingInVector(tempRingInData[vect_num]);
+                ringInOutSwitch = 1;
```
Comment (jordenh) [inline PMMain/Midas/MidasGUI/MidasThread.cpp]: Recall we had talked about getting an actual handle to the ringVector and pushing back individual data items into them, rather than setting the whole vector each time. This is still needed to be implemented.

File: PMMain/Midas/MidasGUI/MidasThread.cpp [hunk 6]
```diff
@@ -41,3 +42,97 @@ void MidasThread::threadEmitStateString(std::string str)
...
+            }
+            else if (ringInOutSwitch == 1)
+            {
+                std::cout << "Ring Out Vector!!!!!\n";
+                ringTemp.setRingOutVector(tempRingOutData[vect_num]);
+
```
Comment (jordenh) [inline PMMain/Midas/MidasGUI/MidasThread.cpp]: Ditto.

File: PMMain/Midas/MidasGUI/MidasThread.cpp [hunk 7]
```diff
@@ -41,3 +42,97 @@ void MidasThread::threadEmitStateString(std::string str)
 {
     emit emitStateString(str);
 }
+
+
+std::vector<ringData> MidasThread::readKeyboardSetupFile()
+{
+    
+    std::ifstream readRingData("C:/Users/Owen/Documents/keyboardData.txt");
```
Comment (lordy1992) [inline PMMain/Midas/MidasGUI/MidasThread.cpp]: We shouldn't hardcode the file path in this method. It should probably be passed as a parameter. Also, we shouldn't commit changes (or, at least, try and push changes) that have a specific absolute path. Always deal with relative paths.

File: PMMain/Midas/MidasGUI/MidasThread.h [hunk 8]
```diff
@@ -21,7 +25,7 @@ class MidasThread : public QThread
 
     void threadEmitString(std::string str);
     void threadEmitStateString(std::string str);
-
+    std::vector<ringData> readKeyboardSetupFile();
 private:
```
Comment (jordenh) [inline PMMain/Midas/MidasGUI/MidasThread.h]: As per comment above, this function should be elsewhere (maybe even it's own class "KeyboardSettingsReader" or something)

File: PMMain/Midas/MidasGUI/MidasThread.cpp [hunk 9]
```diff
@@ -41,3 +42,97 @@ void MidasThread::threadEmitStateString(std::string str)
...
+    std::string temp;
+    int vect_num = 0;
+    int ringInOutSwitch = 0; //0 = ringIn , 1 = ringOut
```
Comment (lordy1992) [inline PMMain/Midas/MidasGUI/MidasThread.cpp]: If 0 is supposed to represent ringIn and 1 is supposed to represent ringOut, you should define constants or create an enum instead of writing the intended meaning in a comment. It's a lot clearer to a reader of the code.

File: PMMain/Midas/MidasGUI/RingData.cpp [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+
+ringData::~ringData()
+{
+}
+
+std::vector<ringData::keyboard_value> ringData::getRingInVector()
+{
+    return*ringInVector;
+}
```
Comment (jordenh) [inline PMMain/Midas/MidasGUI/RingData.cpp]: As in other comments, this should return the pointer, not the dereference object or else it can't be modified outside this class without a full "setRingInVector" callback (waste of time/space). The mutators don't really need to exist by this definition.

Comment (jordenh) [inline PMMain/Midas/MidasGUI/RingData.cpp]: Also, the function name could be like getRingInVectorHandle()

File: PMMain/Midas/MidasGUI/RingData.cpp [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+
+ringData::ringData()
+{
+}
+
+
+ringData::~ringData()
+{
+}
```
Comment (jordenh) [inline PMMain/Midas/MidasGUI/RingData.cpp]: Pointers need to be dealt with in destructor.

File: PMMain/Midas/MidasGUI/RingData.h [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    void Dispose_object()
+    {
+        delete this;
+    }
+
+    struct keyboard_value
+    {
+        char main, hold;
+        keyboard_value(char mainVal = '\0', char holdVal = '\0') : main(mainVal), hold(holdVal) { }
+    };
```
Comment (jordenh) [inline PMMain/Midas/MidasGUI/RingData.h]: mainVal shouldnt have a default value. Rather, it should be mandatory to populate it.

File: PMMain/Midas/MidasGUI/SCDDigester.cpp [hunk 13]
```diff
@@ -131,7 +138,11 @@ void SCDDigester::digestKeyboardData(commandData nextCommand)
             EXACT same thing as select except that the HOLD character is used, rather than the regular character.
             // in the discusion of RingData, this would correspond to the "*Hold vectors"
             */
-
+            if (kybdGUISel % 4 == 0)
+                kybrdCtrl->setKeyChar(kybrdRingData[kybdGUISel].getRingInVector()[scdHandle->getKeySelectAngle() /  (360/kybrdRingData[kybdGUISel].getRingInVector().size())].hold);
+            else
+                kybrdCtrl->setKeyChar(kybrdRingData[kybdGUISel].getRingOutVector()[scdHandle->getKeySelectAngle() / (360/kybrdRingData[kybdGUISel].getRingOutVector().size())].hold);
```
Comment (jordenh) [inline PMMain/Midas/MidasGUI/SCDDigester.cpp]: This logic definitely needs to be modularized. Too hard to follow in single lines.

File: PMMain/Midas/MidasGUI/SharedCommandData.cpp [hunk 14]
```diff
@@ -219,4 +308,30 @@ void SharedCommandData::extractPoint(boost::any value)
...
+        ori_data orientation = boost::any_cast<ori_data> (value);
+        setMyoOrientation(orientation);
+    }
+}
+
+void SharedCommandData::extractKeySelectAngle(int angle)
+{
+
+    setKeySelectAngle(angle);
+
```
Comment (jordenh) [inline PMMain/Midas/MidasGUI/SharedCommandData.cpp]: just for consistency, can probably take out white spaces

File: PMMain/Midas/MidasGUI/SharedCommandData.h [hunk 15]
```diff
@@ -8,7 +8,8 @@
 
 #define COMMAND_INPUT "command"
 #define VELOCITY_INPUT "velocity"
-
+#define ORIENTATION_INPUT "orientation"
+#define ANGLE_INPUT "angle"
 /**
```
Comment (jordenh) [inline PMMain/Midas/MidasGUI/SharedCommandData.h]: Is ANGLE_INPUT ever used? If not, shouldnt be here

File: PMMain/Midas/MidasGUI/MidasThread.cpp [hunk 16]
```diff
@@ -41,3 +42,97 @@ void MidasThread::threadEmitStateString(std::string str)
...
+{
+    
+    std::ifstream readRingData("C:/Users/Owen/Documents/keyboardData.txt");
+    std::string temp;
+    int vect_num = 0;
+    int ringInOutSwitch = 0; //0 = ringIn , 1 = ringOut 
+    bool holdkey_check = false;
+
+    ringData::keyboard_value key_temp; //lvl1
+    std::vector<ringData::keyboard_value> tempRingInData[100]; //lvl2
```
Comment (lordy1992) [inline PMMain/Midas/MidasGUI/MidasThread.cpp]: Why do you need to have 100 vectors for the inner ring? What does this represent? I feel like this should be a different data structure; 100 vectors seems excessive.

File: PMMain/Midas/MidasGUI/SharedCommandData.h (hunk 15 above)
Comment (jordenh) [inline PMMain/Midas/MidasGUI/SharedCommandData.h]: Oh it seems that it is. Why is angle and orientation data being passed?

File: PMMain/Midas/MidasGUI/MidasThread.cpp [hunk 17]
```diff
@@ -41,3 +42,97 @@ void MidasThread::threadEmitStateString(std::string str)
...
+    std::vector<ringData::keyboard_value> tempRingOutData[100]; //lvl2
+    ringData ringTemp; //lvl3
+    std::vector<ringData> keyboardData; //lvl4
```
Comment (lordy1992) [inline PMMain/Midas/MidasGUI/MidasThread.cpp]: The comments on each line aren't instructive. What does lvl1 - lvl4 mean?

File: PMMain/Midas/MidasGUI/keyboardData.txt [hunk 18]
```diff
@@ -0,0 +1,4 @@
+aA bB cC dD eE fF gG hH iI jJ kK lL mM
+nN oO pP qQ rR sS tT uU vV wW xX yY zZ
+`~ 1! 2@ 3# 4$ 5% 6^ 7& 8* 9( 0) -_ =+
+[{ ]} ;: '" \| ,< .> |\ /? th is te nd
```
Comment (jordenh) [inline PMMain/Midas/MidasGUI/keyboardData.txt]: what is the purpose of "th is te nd" ? Shouldnt be needed since theres no requirement for sizes of 13.

File: PMMain/Midas/MidasGUI/kybrdCtrl.cpp [hunk 19]
```diff
@@ -1,6 +1,7 @@
 #include "kybrdCtrl.h"
 #include <iostream>
-
+#include <string>
+#include <fstream>
 KybrdCtrl::KybrdCtrl()
```
Comment (jordenh) [inline PMMain/Midas/MidasGUI/kybrdCtrl.cpp]: kybrdCtrl.cpp/h can be reverted as I dont see any actual content added, correct?

File: PMMain/Midas/MidasGUI/MidasThread.cpp [hunk 20]
```diff
@@ -41,3 +42,97 @@ void MidasThread::threadEmitStateString(std::string str)
...
+
+
+    //readRingData.open();
+    if (readRingData.is_open())
+    {
+        while (!readRingData.eof())
+        {
+            getline(readRingData, temp);
+            char* temp_char = (char*)malloc(sizeof(char)*(temp.length() + 1));
+            strcpy(temp_char, temp.c_str());
```
Comment (lordy1992) [inline PMMain/Midas/MidasGUI/MidasThread.cpp]: Why not just use C++ strings? No need to malloc or anything, just use the C++ string you read from the file.

File: PMMain/Midas/MidasGUI/MidasThread.cpp [hunk 21]
```diff
@@ -41,3 +42,97 @@ void MidasThread::threadEmitStateString(std::string str)
...
+
+            for (int i = 0; i < temp.length(); i++)
+            {
+
+                if (temp_char[i] != char(32) && holdkey_check == false)
```
Comment (lordy1992) [inline PMMain/Midas/MidasGUI/MidasThread.cpp]: I don't know what you're trying to do in the first part of this boolean expression. But if you're trying to tell if temp_char[i] is an ASCII character, this isn't doing what you think it's doing.

File: PMMain/Midas/MidasGUI/MidasThread.cpp [hunk 22]
```diff
@@ -41,3 +42,97 @@ void MidasThread::threadEmitStateString(std::string str)
...
+                {
+
+                    key_temp.main = temp_char[i];
+
+
+                    if (temp_char[i + 1] == char(32))
```
Comment (lordy1992) [inline PMMain/Midas/MidasGUI/MidasThread.cpp]: Again, not going to do what you expect. This is just checking to see if temp_char[i+1] is equal to 32 (cast to a char). Is that what you meant?

File: PMMain/Midas/MidasGUI/MidasThread.cpp [hunk 23]
```diff
@@ -41,3 +42,97 @@ void MidasThread::threadEmitStateString(std::string str)
...
+                    {
+                        holdkey_check = false;
+
+                        if (ringInOutSwitch == 0)
+                            tempRingInData[vect_num].push_back(key_temp);
```
Comment (lordy1992) [inline PMMain/Midas/MidasGUI/MidasThread.cpp]: If you're doing an if with an else as well, I recommend using braces. I'm okay with one-liners if there is no 'else' part, like:

if (condition) expression;

But it's prone to error if you do it like this, with an else as well.

File: PMMain/Midas/MidasGUI/MidasThread.cpp (hunk 21 above)
Comment (owen927) [inline PMMain/Midas/MidasGUI/MidasThread.cpp]: char(32) is space, its checking if the current char is a space

File: PMMain/Midas/MidasGUI/kybrdCtrl.cpp (hunk 19 above)
Comment (owen927) [inline PMMain/Midas/MidasGUI/kybrdCtrl.cpp]: yeah sorry i probably made a misclick here

File: PMMain/Midas/MidasGUI/RingData.cpp [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+
+std::vector<ringData::keyboard_value> ringData::getRingInVector()
+{
+    return*ringInVector;
```
Comment (lordy1992) [inline PMMain/Midas/MidasGUI/RingData.cpp]: Put a space between return and *ringInVector.

File: PMMain/Midas/MidasGUI/RingData.cpp [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+std::vector<ringData::keyboard_value> ringData::getRingOutVector()
+{
+    return*ringOutVector;
```
Comment (lordy1992) [inline PMMain/Midas/MidasGUI/RingData.cpp]: Same as above

File: PMMain/Midas/MidasGUI/keyboardData.txt (hunk 18 above)
Comment (owen927) [inline PMMain/Midas/MidasGUI/keyboardData.txt]: sorry OCD, you are correct there is no requirement for sizes of 13

File: PMMain/Midas/MidasGUI/RingData.cpp [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+void ringData::setRingInVector(std::vector<keyboard_value> &ringInVal)
+{
+    ringInVector = &ringInVal;
```
Comment (lordy1992) [inline PMMain/Midas/MidasGUI/RingData.cpp]: This seems a bit strange. Why are you storing a handle to what the caller passes in instead of just copying the data?

File: PMMain/Midas/MidasGUI/RingData.h [hunk 27]
```diff
@@ -0,0 +1,34 @@
+#include <vector>
+#pragma once
+class ringData
+{
+public:
+    ringData();
+    ~ringData();
+
+    void Dispose_object()
```
Comment (lordy1992) [inline PMMain/Midas/MidasGUI/RingData.h]: Your naming convention is inconsistent. Some of your methods use camelcase (i.e. thisIsAFunction() ), and some use the underscores (this_is_a_function()). We've been using camelcase for everything else, so you should change your names to reflect that.

File: PMMain/Midas/MidasGUI/SCDDigester.cpp [hunk 28]
```diff
@@ -115,10 +117,15 @@ void SCDDigester::digestKeyboardData(commandData nextCommand)
             //threadHandle->UpdateGUIToChangeWheels TODO
             scdHandle->setKybdGuiSel(kybdGUISel);
             break;
+
         case kybdGUICmds::SELECT:
+            if (kybdGUISel % 4 == 0)
+                kybrdCtrl->setKeyChar(kybrdRingData[kybdGUISel].getRingInVector()[scdHandle->getKeySelectAngle() / (360 / kybrdRingData[kybdGUISel].getRingInVector().size())].main);
```
Comment (lordy1992) [inline PMMain/Midas/MidasGUI/SCDDigester.cpp]: You should probably split this up into more than one line. If anything goes wrong, it will be hard to tell what calculation is incorrect. Also, it may be worth checking that the indices are valid.

File: PMMain/Midas/MidasGUI/SCDDigester.cpp [hunk 29]
```diff
@@ -131,7 +138,11 @@ void SCDDigester::digestKeyboardData(commandData nextCommand)
             EXACT same thing as select except that the HOLD character is used, rather than the regular character.
             // in the discusion of RingData, this would correspond to the "*Hold vectors"
             */
-
+            if (kybdGUISel % 4 == 0)
+                kybrdCtrl->setKeyChar(kybrdRingData[kybdGUISel].getRingInVector()[scdHandle->getKeySelectAngle() /  (360/kybrdRingData[kybdGUISel].getRingInVector().size())].hold);
+            else
+                kybrdCtrl->setKeyChar(kybrdRingData[kybdGUISel].getRingOutVector()[scdHandle->getKeySelectAngle() / (360/kybrdRingData[kybdGUISel].getRingOutVector().size())].hold);
+            kybrdCtrl->sendData();
```
Comment (lordy1992) [inline PMMain/Midas/MidasGUI/SCDDigester.cpp]: Same as above for these lines

File: PMMain/Midas/MidasGUI/MidasThread.cpp (hunk 21 above)
Comment (lordy1992) [inline PMMain/Midas/MidasGUI/MidasThread.cpp]: Hmm... If you want to check for space use ' ' instead. And you're only allowing one space between?
